src/idpi/product.py

```diff
@@ -0,0 +1,58 @@
+"""Product base classes."""
+
+# Standard library
+import typing
+from abc import ABCMeta
+from abc import abstractmethod
+from itertools import accumulate
+
+# Third-party
+import dask
+
+
+class Register:
+    """Register methods that are (dask) delayed for caching."""
+
+    def __init__(self, delay: bool = False):
+        self.regdict: dict[str, typing.Any] = {}
+        self._delayed = dask.delayed if delay else lambda x: x
+
+    def reg(self, fn, *arg):
+        key = list(
+            accumulate([hash(id(x)) for x in arg], lambda x, acc: hash(x ^ acc))
+        )[-1]
+
+        return self.regdict.setdefault(key, self._delayed(fn)(*arg))
+
+
+class Product(metaclass=ABCMeta):
+    """Base class for products."""
+
+    def __init__(
+        self,
+        input_fields: list[str],
+        reg: Register | None = None,
+        delay_entire_product: bool = False,
+    ):
+        self._input_fields = input_fields
+        if not reg:
+            self.reg = Register()
+        else:
+            self.reg = reg
+        self._base_delayed = dask.delayed if delay_entire_product else lambda x: x
+
+    # avoid a possible override from inheriting classes
+    @property
+    def delay_entire_product(self):
+        return self._base_delayed
```

Comment:
Can you explain why this is needed?

Reply:
The main idea is that the base class will control the delay of entire products run, i.e. the situation where the entire product processing goes into one single task. I think this should cover most of the cases. 
However sometimes, we might want to have more fine grained control of the tasking, in case various products are sharing some intermediate computations. 
In that case, we would like to not introduce the delay at the product level (that is this property should be set to False), and the child class has the responsibility of delaying the various functions and operations in a more fine grained controled manner. 
The child class has another delay parameter which allows to activate/deactivate delaying tasks. So that we can also run products without dask delaying any computation. 
https://github.com/MeteoSwiss-APN/idpi-products/pull/6/files#diff-832bf11faf4adb36354b27b36199692f18199af6252edff013a54c9db72fec79R17

So the main intention here was to make this delay_entire_product a property, to avoid that the child classes overwrite the value of it, since there are similar attributes in childern. So that this can only be set at the init. 
Does it answer the question, does it make sense?

Reply:
Following our discussion, it seems like the `delay_entire_product` property is not intended to be used other than to indicate that the `_base_delay` should not be overridden in the child classes